### PR TITLE
Analytical_Engine: Handle nulls

### LINKED ISSUE #2472: Analytical_Engine - handle nulls
For the following methods:

 - [x] BH.Engine.Analytical.Convert.ToDataVizObjects(Graph graph)
 - [x] BH.Engine.Analytical.Convert.ToDotFormat(Graph graph, String shape, Int32 fontsize)
 - [x] BH.Engine.Analytical.Convert.ToRelation(SourcesDependencyFragment dependency, Guid owningEntity)
 - [x] BH.Engine.Analytical.Convert.ToRelation(TargetsDependencyFragment dependency, Guid owningEntity)
 - [x] BH.Engine.Analytical.Convert.ToRelation(IBHoMObject obj)
 - [x] BH.oM.Analytical.Elements.Relation.Relation(IBHoMObject source, IBHoMObject target, Graph subgraph, Double weight, ICurve curve)
 - [x] BH.Engine.Analytical.Modify.RemoveEntity(Graph graph, IBHoMObject entityToRemove)
 - [x] BH.Engine.Analytical.Modify.RemoveEntity(Graph graph, Guid entityToRemove)
 - [x] BH.Engine.Analytical.Modify.RemoveIsolatedEntities(Graph graph)
 - [x] BH.Engine.Analytical.Modify.Reverse(Relation relation)
 - [x] BH.Engine.Analytical.Modify.Reverse(Graph graph)
 - [x] BH.Engine.Analytical.Modify.SetGeometry(INode node, Point point)
 - [x] BH.Engine.Analytical.Modify.SetGeometry(IEdge edge, ICurve curve)
 - [x] BH.Engine.Analytical.Modify.SetGeometry(ISurface anaSurface, ISurface geoSurface)
 - [x] BH.Engine.Analytical.Modify.SetOutlineElements1D(IRegion region, IEnumerable<IElement1D> outlineElements)
 - [x] BH.Engine.Analytical.Modify.UniqueEntityNames(Graph graph)
 - [x] BH.Engine.Analytical.Query.Geometry(Graph graph)
 - [x] BH.Engine.Analytical.Query.Adjacency(Graph graph, RelationDirection relationDirection)
 - [x] BH.Engine.Analytical.Query.EntityNeighbourhood(Graph graph, RelationDirection relationDirection)
 - [x] BH.Engine.Analytical.Query.EntityNeighbourhood(Graph graph, IBHoMObject entity, Int32 maximumDepth, RelationDirection relationDirection)
 - [x] BH.Engine.Analytical.Query.AStarShortestPath(Graph graph, IBHoMObject start, IBHoMObject end)
 - [x] BH.Engine.Analytical.Query.AStarShortestPath(Graph graph, Guid start, Guid end)
 - [x] BH.Engine.Analytical.Query.Depth(Graph graph, Guid startEntity)
 - [x] BH.Engine.Analytical.Query.Destinations(Graph graph, Guid entity)
 - [x] BH.Engine.Analytical.Query.Destinations(Graph graph, IBHoMObject entity)
 - [x] BH.Engine.Analytical.Query.DijkstraShortestPath(Graph graph, IBHoMObject start, IBHoMObject end)
 - [x] BH.Engine.Analytical.Query.DijkstraShortestPath(Graph graph, Guid start, Guid end)
 - [x] BH.Engine.Analytical.Query.Entities(Graph graph)
 - [x] BH.Engine.Analytical.Query.Entity(Graph graph, String entityName)
 - [x] BH.Engine.Analytical.Query.Entity(Graph graph, Guid entityGuid)
 - [x] BH.Engine.Analytical.Query.FilterEntities(Graph graph, Type typeFilter)
 - [x] BH.Engine.Analytical.Query.FilterRelations(Graph graph, Type typeFilter)
 - [x] BH.Engine.Analytical.Query.Incoming(Graph graph, Guid entity)
 - [x] BH.Engine.Analytical.Query.Incoming(Graph graph, IBHoMObject entity)
 - [x] BH.Engine.Analytical.Query.Relation(Graph graph, IBHoMObject source, IBHoMObject target, RelationDirection relationDirection)
 - [x] BH.Engine.Analytical.Query.Relation(Graph graph, Guid source, Guid target, RelationDirection relationDirection)
 - [x] BH.Engine.Analytical.Query.RelationLength(Graph graph, IRelation relation)
 - [x] BH.Engine.Analytical.Query.NotSinks(Graph graph)
 - [x] BH.Engine.Analytical.Query.Sinks(Graph graph)
 - [x] BH.Engine.Analytical.Query.Sources(Graph graph)
 - [x] BH.Engine.Analytical.Query.SubGraphs(Graph graph, RelationDirection relationDirection)
 - [x] BH.Engine.Analytical.Query.IsolatedEntities(Graph graph)
 - [x] BH.Engine.Analytical.Query.OutlineElements1D(IRegion region)

### PR DESCRIPTION
**Pull request by FraserGreenroyd**
### Issues addressed by this PR
Fixes #2472

### CODE REVIEW

**Comment by FraserGreenroyd** [comment]
@BHoMBot check null-handling

**Comment by FraserGreenroyd** [comment]
@BHoMBot check null-handling

**Comment by FraserGreenroyd** [comment]
@BHoMBot check required

**Review by pawelbaran — APPROVED** [review]
Similarly to #2465, I have just pushed a commit changing the return object from null reference to `null`. Besides that I have no further objections outside of what I have mentioned in my previous comments in previous PRs of this type.

As mentioned in #2465, I am a bit concerned about the creep that these changes may possibly introduce, but it seems that the only way to check it is to merge and see how many weird behaviours bubble up.

**Comment by pawelbaran** [comment]
@BHoMBot check null-handling

File: Analytical_Engine/Query/Graph/EntityNeighbourhood.cs [hunk 1]
```diff
@@ -46,6 +46,12 @@ public static partial class Query
         [Output("graphs", "The collection of sub Graphs found in the input Graph.")]
         public static List<Graph> EntityNeighbourhood(this Graph graph, RelationDirection relationDirection = RelationDirection.Forwards)
         {
+            if (graph == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("Cannot query the entity neighbourhood of a null graph.");
+                return null;
```
Comment (pawelbaran) [inline Analytical_Engine/Query/Graph/EntityNeighbourhood.cs:52]: Why not return a list to follow the pattern?
```suggestion
                return new List<Graph>();
```

File: Analytical_Engine/Query/Graph/FilterEntities.cs [hunk 2]
```diff
@@ -46,6 +46,12 @@ public static partial class Query
         [Output("filter entities", "Entity Dictionary containing the filtered entities.")]
         public static Dictionary<Guid, IBHoMObject> FilterEntities(this Graph graph, Type typeFilter)
         {
+            if (graph == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("Cannot filter the entities of a null graph.");
+                return null;
```
Comment (pawelbaran) [inline Analytical_Engine/Query/Graph/FilterEntities.cs:52]: Why not return an empty collection to follow the pattern?
```suggestion
                return new Dictionary<Guid, IBHoMObject>();
```

File: Analytical_Engine/Query/Graph/FilterRelations.cs [hunk 3]
```diff
@@ -45,6 +45,12 @@ public static partial class Query
         [Output("filtered relations", "Collection of IRelations filtered from the Graph.")]
         public static List<IRelation> FilterRelations(this Graph graph, Type typeFilter)
         {
+            if (graph == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("Cannot filter the relations of a null graph.");
+                return null;
```
Comment (pawelbaran) [inline Analytical_Engine/Query/Graph/FilterRelations.cs:51]: Why not return an empty collection to follow the pattern?
```suggestion
                return new List<IRelation>();
```

File: Analytical_Engine/Query/Graph/IsolatedEntities.cs [hunk 4]
```diff
@@ -43,6 +43,12 @@ public static partial class Query
         [Output("isolated entities", "The collection of entity Guids that are isolated.")]
         public static List<Guid> IsolatedEntities(this Graph graph)
         {
+            if(graph == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("Cannot query the isolated entities of a null graph.");
+                return null;
```
Comment (pawelbaran) [inline Analytical_Engine/Query/Graph/IsolatedEntities.cs:49]: Why not return an empty collection to follow the pattern?
```suggestion
                return new List<Guid>();
```

File: Analytical_Engine/Query/Graph/Relation.cs [hunk 5]
```diff
@@ -46,6 +46,24 @@ public static partial class Query
         [Output("relations", "The collection of matching IRelations.")]
         public static List<IRelation> Relation(this Graph graph, IBHoMObject source, IBHoMObject target, RelationDirection relationDirection = RelationDirection.Forwards)
         {
+            if (graph == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("Cannot query the relations of a null graph.");
+                return null;
```
Comment (pawelbaran) [inline Analytical_Engine/Query/Graph/Relation.cs:52]: Why not return an empty collection to follow the pattern?
```suggestion
                return new List<IRelation>();
```

File: Analytical_Engine/Query/Graph/Relation.cs [hunk 6]
```diff
@@ -46,6 +46,24 @@ public static partial class Query
...
+            }
+
+            if (source == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("Cannot query the relations of a graph when the source is null.");
+                return null;
```
Comment (pawelbaran) [inline Analytical_Engine/Query/Graph/Relation.cs:58]: Why not return an empty collection to follow the pattern?
```suggestion
                return new List<IRelation>();
```

File: Analytical_Engine/Query/Graph/Relation.cs [hunk 7]
```diff
@@ -46,6 +46,24 @@ public static partial class Query
...
+            }
+
+            if (target == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("Cannot query the relations of a graph when the target is null.");
+                return null;
```
Comment (pawelbaran) [inline Analytical_Engine/Query/Graph/Relation.cs:64]: Why not return an empty collection to follow the pattern?
```suggestion
                return new List<IRelation>();
```

File: Analytical_Engine/Query/Graph/Relation.cs [hunk 8]
```diff
@@ -75,6 +93,12 @@ public static List<IRelation> Relation(this Graph graph, IBHoMObject source, IBH
         [Output("relations", "The collection of matching IRelation Guids.")]
         public static List<Guid> Relation(this Graph graph, Guid source, Guid target, RelationDirection relationDirection = RelationDirection.Forwards)
         {
+            if (graph == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("Cannot query the relations of a null graph.");
+                return null;
```
Comment (pawelbaran) [inline Analytical_Engine/Query/Graph/Relation.cs:99]: Why not return an empty collection to follow the pattern?
```suggestion
                return new List<Guid>();
```

File: Analytical_Engine/Query/Graph/Adjacency.cs [hunk 9]
```diff
@@ -44,6 +44,12 @@ public static partial class Query
         [Output("adjacency", "The Dictionary where the keys are entities and the values are the collection of adjacent entities.")]
         public static Dictionary<Guid, List<Guid>> Adjacency(this Graph graph, RelationDirection relationDirection = RelationDirection.Forwards)
         {
+            if(graph == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("Cannot query the adjacency of a null graph.");
+                return null;
```
Comment (pawelbaran) [inline Analytical_Engine/Query/Graph/Adjacency.cs:50]: Why not return an empty collection to follow the pattern?
```suggestion
                return new Dictionary<Guid, List<Guid>>();
```

**Review by pawelbaran — CHANGES_REQUESTED** [review]
Changed my mind sorry - please see the code comments.

**Comment by FraserGreenroyd** [comment]
@BHoMBot check required

File: Analytical_Engine/Query/Graph/Depth.cs [hunk 10]
```diff
@@ -93,6 +93,12 @@ public static partial class Query
         [Output("depths", "A Dictionary of the depths of the entities in the Graph.")]
         public static Dictionary<Guid, int> Depth(this Graph graph, Guid startEntity)
         {
+            if(graph == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("Cannot query the depth of a null graph.");
+                return null;
```
Comment (pawelbaran) [inline Analytical_Engine/Query/Graph/Depth.cs:99]: To follow a single pattern:
```suggestion
                return new Dictionary<Guid, int>();
```

**Review by pawelbaran — CHANGES_REQUESTED** [review]

**Review by pawelbaran — APPROVED** [review]
I have no objections outside of what I have mentioned in my previous comments in previous PRs of this type.

As mentioned in #2465, I am a bit concerned about the creep that these changes may possibly introduce, but it seems that the only way to check it is to merge and see how many weird behaviours bubble up.

**Comment by pawelbaran** [comment]
@BHoMBot check required

**Comment by FraserGreenroyd** [comment]
@BHoMBot check copyright-compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check dataset-compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot this is a CI/CD instruction. I am authorising dispensation to be granted on check ref. 2411206415

**Pull request merged**